### DISCUSSION TITLE
How to release memory when window close

### DISCUSSION THREAD

**Discussion by Coloryr** [AvaloniaUI/Avalonia#10111] [Q&A] [answered]
I open and close window many times.
```
public static void ShowSetting()
    {
        if (SettingWindow != null)
        {
            SettingWindow.Activate();
        }
        else
        {
            SettingWindow = new();
            SettingWindow.Show();
        }
    }

 public static void Test()
    {
        for (int a = 0; a < 100; a++)
        {
            ShowSetting();
            SettingWindow.Close();
        }
    }
```

```
private void SettingWindow_Closed(object? sender, EventArgs e)
    {
        App.PicUpdate -= Update;

        App.SettingWindow = null;

        tab1.SetWindow(null);
        tab2.SetWindow(null);
        tab3.SetWindow(null);
        tab4.SetWindow(null);
        tab5.SetWindow(null);
        tab6.SetWindow(null);

        tab1 = null;
        tab2 = null;
        tab3 = null;
        tab4 = null;
        tab5 = null;
        tab6 = null;

        content1 = null;
        content2 = null;
    }
```
but they are not release, it use a lot memory, how to release that.

![{G%(GK 4Y VHM7QRAGUQFGP](https://user-images.githubusercontent.com/26276037/215250266-be91f08d-3307-4aad-a165-7bde677d1c6d.png)

**Comment by maxkatz6** [comment, marked as answer]
It looks like you have a memory leak somewhere.
Avalonia Windows are automatically released when they are closed and not used anywhere anymore.
With tools like dotMemory or VS memory profiler you can find what's holding these objects.

**Comment by Coloryr** [comment]
It release when `get snapshot` ???

![屏幕截图 2023-01-28 165650](https://user-images.githubusercontent.com/26276037/215257076-7c4ad217-4457-40c7-ae63-3e1dd1806bfb.png)

**Reply by maxkatz6** [reply]
Yes, it got dotMemory forced garbage collector to remove these objects from memory.
If windows were removed from memory after that, then you didn't have any memory leak, so it's good. You just need to wait when garbage collector does its job automatically.
See these conditions for the garbage collector: https://learn.microsoft.com/en-us/dotnet/standard/garbage-collection/fundamentals#conditions-for-a-garbage-collection

**Reply by Coloryr** [reply]
> Yes, it got dotMemory forced garbage collector to remove these objects from memory. If windows were removed from memory after that, then you didn't have any memory leak, so it's good. You just need to wait when garbage collector does its job automatically. See these conditions for the garbage collector: https://learn.microsoft.com/en-us/dotnet/standard/garbage-collection/fundamentals#conditions-for-a-garbage-collection

Thanks, i change the `GC Type` it will collect in close 
```
GCSettings.LatencyMode = GCLatencyMode.LowLatency;
```

**Comment by Coloryr** [comment]
![屏幕截图 2023-01-28 165650](https://user-images.githubusercontent.com/26276037/215257225-3ed3e9ab-c6db-45be-811a-b269f6b26f01.png)

**Comment by jetspiking** [comment]
@Coloryr Using .NET the memory is not always released instantly by the GC after it's lifecycle (unlike RAII). Provided you didn't use any unsafe code, lots of static references, threading, it seems unlikely to me that a memory leak could be the issue. In your latest image I see the memory usage of `SettingWindow` is `2`. You mentioned earlier you adjusted the `GCSettings.LatencyMode`, did this fix the issues for you?
